jupyterhub/tests/test_scopes.py

```diff
@@ -760,3 +760,66 @@ async def test_resolve_token_permissions(
     token_retained_scopes = get_scopes_for(orm_api_token)
 
     assert token_retained_scopes == intersection_scopes
+
+
+@pytest.mark.parametrize(
+    "left, right, expected, should_warn",
+    [
+        (set(), set(), set(), False),
+        (set(), set(["users"]), set(), False),
+        # no warning if users and groups only on the same side
+        (
+            set(["users!user=x", "users!group=y"]),
+            set([]),
+            set([]),
+            False,
+        ),
+        # no warning if users are on both sizes
+        (
+            set(["users!user=x", "users!user=y", "users!group=y"]),
+            set(["users!user=x"]),
+            set(["users!user=x"]),
+            False,
+        ),
+        # no warning if users and groups are both defined
+        # on both sides
+        (
+            set(["users!user=x", "users!group=y"]),
+            set(["users!user=x", "users!group=y", "users!user=z"]),
+            set(["users!user=x", "users!group=y"]),
+            False,
+        ),
+        # warn if there's a user on one side and a group on the other
+        # which *may* intersect
+        (
+            set(["users!group=y", "users!user=z"]),
+            set(["users!user=x"]),
+            set([]),
+            True,
+        ),
+        # and the reverse, just in case
+        (
+            set(["users!user=x"]),
+            set(["users!group=y"]),
+            set([]),
+            True,
+        ),
+        # resolves server under user, without warning
```

Comment:
Maybe adding another one where it resolves to an empty set? Probably not needed as the code does check properly for the username match.

Reply:
Added, thanks!